File: Dopamine.ControlsModule/ViewModels/SearchControlViewModel.vb
Imports Microsoft.Practices.Prism.Mvvm
Imports Dopamine.Common.Services.Search

Namespace ViewModels
    Public Class SearchControlViewModel
        Inherits BindableBase

#Region "Variables"
        Private mSearchText As String
        Private mSearchService As ISearchService
#End Region

#Region "Properties"
        Public Property SearchText() As String
            Get
                Return mSearchText
            End Get
            Set(ByVal value As String)
                SetProperty(Of String)(mSearchText, value)
                mSearchService.SearchText = value
            End Set
        End Property
#End Region

#Region "Construction"
        Public Sub New(iSearchService As ISearchService)
            mSearchService = iSearchService

            AddHandler mSearchService.DoSearch, Sub() Me.UpdateSearchText()

            Me.UpdateSearchText()
        End Sub
#End Region

#Region "Private"
        Private Sub UpdateSearchText()
            mSearchText = mSearchService.SearchText
            OnPropertyChanged(Function() Me.SearchText)
        End Sub
#End Region
    End Class
End Namespace

File: Dopamine.ControlsModule/ViewModels/SearchControlViewModel.cs
using Dopamine.Common.Services.Search;
using Microsoft.Practices.Prism.Mvvm;

namespace Dopamine.ControlsModule.ViewModels
{
    public class SearchControlViewModel : BindableBase
    {
        #region Variables
        private string searchText;
        private ISearchService searchService;
        #endregion

        #region Properties
        public string SearchText
        {
            get { return this.searchText; }
            set
            {
                SetProperty<string>(ref this.searchText, value);
                this.searchService.SearchText = value;
            }
        }
        #endregion

        #region Construction
        public SearchControlViewModel(ISearchService searchService)
        {
            this.searchService = searchService;

            this.searchService.DoSearch += (searchText) => this.UpdateSearchText();

            this.UpdateSearchText();
        }
        #endregion

        #region Private
        private void UpdateSearchText()
        {
            this.searchText = this.searchService.SearchText;
            OnPropertyChanged(() => this.SearchText);
        }
        #endregion
    }
}
